### PR TITLE
bpo-31718: Raise ValueError instead of SystemError when calling methods of uninitialized io.IncrementalNewlineDecoder objects

### PR DESCRIPTION
**Pull request by orenmn**
In addition:
* Add tests to `test_io` to verify that indeed `ValueError` is raised instead of `SystemError`.
* While we are here, fix refleaks in ``io.IncrementalNewlineDecoder.__init__()``.

https://bugs.python.org/issue31718

### CODE REVIEW

File: Modules/_io/textio.c [hunk 1]
```diff
@@ -309,8 +311,9 @@ _PyIncrementalNewlineDecoder_decode(PyObject *myself,
     nldecoder_object *self = (nldecoder_object *) myself;
 
     if (self->decoder == NULL) {
-        PyErr_SetString(PyExc_ValueError,
-                        "IncrementalNewlineDecoder.__init__ not called");
+        PyErr_Format(PyExc_ValueError,
+                     "%.200s.__init__() not called",
```
Comment (serhiy-storchaka) [inline Modules/_io/textio.c]: Revert this change. The problem is that `IncrementalNewlineDecoder.__init__` not called. This can happen if subclass' `__init__` doesn't call `super().__init__`. Subclass's `__init__` can be called, but `self->decoder == NULL` if `IncrementalNewlineDecoder.__init__` not called.

File: Lib/test/test_io.py [hunk 2]
```diff
@@ -3508,7 +3508,13 @@ def _check(dec):
         _check(dec)
 
 class CIncrementalNewlineDecoderTest(IncrementalNewlineDecoderTest):
-    pass
+    def test_uninitialized(self):
```
Comment (serhiy-storchaka) [inline Lib/test/test_io.py:3512]: Add cpython_only? This is an implementation detail, if implement the constructor as `__new__` rather than `__init__`, the following errors will be not raised.

File: Modules/_io/textio.c [hunk 3]
```diff
@@ -248,9 +248,11 @@ _io_IncrementalNewlineDecoder___init___impl(nldecoder_object *self,
                                             PyObject *errors)
 /*[clinic end generated code: output=fbd04d443e764ec2 input=89db6b19c6b126bf]*/
 {
+    Py_XDECREF(self->decoder);
```
Comment (serhiy-storchaka) [inline Modules/_io/textio.c]: There is a race condition, destructor of `self->decoder` can call `IncrementalNewlineDecoder.__init__` . Use Py_XSETREF().

**Comment by orenmn** [comment]
Sorry about messing up so much.

**Review by LuSshkey — APPROVED** [review]

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by serhiy-storchaka** [comment]
LGTM. I'll merge if @benjaminp doesn't have comments.

**Review by benjaminp — COMMENTED** [review]
Two minor things:
1. Most of the _io module classes abstract away the reinitialization logic with a macro `CHECK_INITIALIZED`. It would be nice to do that here.
2. Checking for initialization should be the first thing a method does. In this implementation, `.reset()` can partially reset the object before raising an exception.

**Comment by orenmn** [comment]
Note that i added an `ok` field to `nldecoder_object`. I would change `CHECK_INITIALIZED_DECODER()` to rely on the `decoder` field if you think it is too expensive.
Also, i added `CHECK_INITIALIZED_DECODER()` to the beginning of the getter of `newlines`. I would remove that if you think this is too much of a change to the behavior.

**Comment by serhiy-storchaka** [comment]
Python implementation of `.reset()` can partially reset the object before raising an exception.

I don't understand why the `ok` field is added. `__init__()` can fail only if `PyUnicode_FromString("strict")`, but the `error` field is not used and could be removed (but see bpo-31722). I would set `error = PyUnicode_FromString("strict")` if `error == NULL` before changing any fields (if we want to keep an unused field).

**Comment by orenmn** [comment]
@benjaminp @serhiy-storchaka so should the C implementation allow `reset()` to partially reset the object?

**Review by auvipy — CHANGES_REQUESTED** [review]
rebase is needed

**Comment by csabella** [comment]
@orenmn, please resolve the merge conflicts.  Thank you!

**Comment by orenmn** [comment]
@csabella I am really sorry, but I am not working on CPython these days..

**Comment by csabella** [comment]
@orenmn, no problem and thanks for responding.  I'm going to close this pull request for now so that someone else can continue with it if they are interested.  If you come back to it later on and it's not resolved yet, feel free to reopen it at that time.  Thanks again!

**Comment by serhiy-storchaka** [comment]
It would be better to not close abandoned PRs which fix a bug or implement a valuable enhancement, but recreate them, update and merge.

**Comment by csabella** [comment]
Sorry, but that was my point in closing the abandoned PRs, to allow someone the opportunity to do just that.  If someone becomes aware that a PR exists but isn't being worked on, then it gives that person a chance to pick up the issue and continue it.  Otherwise these PRs are just being left open and no one seems to be aware of them.  I thought I was implementing a good workflow for the older issues, but perhaps it should be discussed more and then formalized?
